arch/x86/compare256_avx512.c

```diff
@@ -0,0 +1,95 @@
+/* compare256_avx512.c -- AVX512 version of compare256
+ * Copyright (C) 2025 Hans Kristian Rosbach
+ * Based on AVX2 implementation by Mika T. Lindqvist
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+#include "zbuild.h"
+#include "zmemory.h"
+#include "deflate.h"
+#include "fallback_builtins.h"
+
+#if defined(X86_AVX512) && defined(HAVE_BUILTIN_CTZLL)
+
+#include <immintrin.h>
+#ifdef _MSC_VER
+#  include <nmmintrin.h>
+#endif
+
+static inline uint32_t compare256_avx512_static(const uint8_t *src0, const uint8_t *src1) {
+    __m512i zmm_src0_4, zmm_src1_4;
+    __m512i zmm_src0_3, zmm_src1_3;
+    __m512i zmm_src0_2, zmm_src1_2;
+    __m512i zmm_src0_1, zmm_src1_1;
+    __m128i xmm_src0_0, xmm_src1_0;
+    uint64_t mask_1, mask_2, mask_3, mask_4;
+    uint32_t mask_0;
+
+    // First do a couple 32byte rounds before increasing to 64bytes, this reduces
+    // the penalty for the short matches, and those are usually the most common ones.
+
+    // 16 bytes
+    xmm_src0_0 = _mm_loadu_si128((__m128i*)src0);
+    xmm_src1_0 = _mm_loadu_si128((__m128i*)src1);
+    mask_0 = (uint32_t)_mm_cmpeq_epu8_mask(xmm_src0_0, xmm_src1_0); // zero-extended to use __builtin_ctz
```

Comment:
I wonder if AVX2 can also benefit from this.

Reply:
The compare with mask integrated?That is a new thing in EVEX encoding (introduced with AVX512) unfortunately, EVEX adds 8 new mask registers in addition to extending vector registers to 512bit and doubling the amount of vector registers.

Reply:
Btw, I think the first two bytes are always guaranteed to be a match.